File: go/array/384_shuffle_array.go
package array

import "math/rand"

type Solution struct {
	data     []int
	original []int
}

func Constructor(nums []int) Solution {
	ori := append([]int(nil), nums...)
	return Solution{
		data:     nums,
		original: ori,
	}
}

func (s *Solution) Reset() []int {

	copy(s.data, s.original)
	return s.data
}

func (s *Solution) Shuffle() []int {
	// 洗牌算法，保证每次取到后的数据都不能再被拿到

	n := len(s.data)
	for i := range s.data {
		// 保证取到的数据 在上次没有取到的数据之后
		j := i + rand.Intn(n-i)
		// 每次随机取之后把取到的数据和 上次的数据换一下
		s.data[i], s.data[j] = s.data[j], s.data[i]
	}

	return s.data
}

File: rust/src/array/384_shuffle_array.rs



struct Solution {
    data: Vec<i32>,
    ori: Vec<i32>,
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Solution {

    fn new(nums: Vec<i32>) -> Self {
        let nums = nums;
        let ori = nums.clone();

        Solution {data: nums, ori: ori }
    }
    
    fn reset(&mut self) -> Vec<i32> {
        self.data = self.ori.clone();
        self.data.clone()
    }
    
    fn shuffle(&mut self) -> Vec<i32> {
        // 洗牌算法
        let n = self.data.len();
        for i in 0..n {

            let j = i + rand::thread_rng().gen_range(0..n-i);
            self.data.swap(i, j);
        }

        self.data.clone()
    }
}